test/view/take_while_view.cpp

```diff
@@ -0,0 +1,92 @@
+// cmcstl2 - A concept-enabled C++ standard library
+//
+//  Copyright Casey Carter
+//  Copyright Eric Niebler
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/caseycarter/cmcstl2
+//
+#include <stl2/view/take_while.hpp>
+#include <stl2/view/iota.hpp>
+#include <list>
+#include <vector>
+#include "../simple_test.hpp"
+
+namespace ranges = __stl2;
+
+namespace view {
+	using namespace ranges::view;
+	using view::ext::take_while;
+} // namespace view
+
+int main()
+{
+	auto rng0 = view::iota(10) | view::take_while([](int i) { return i != 25; });
+	CHECK_EQUAL(rng0, {10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24});
+	static_assert(ranges::View<decltype(rng0)>);
+	// ::models_not<concepts::BoundedView>(aux::copy(rng0));
+	static_assert(ranges::RandomAccessIterator<decltype(rng0.begin())>);
+
+	std::vector<int> vi{0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+	auto rng1 = vi | view::take_while([](int i) { return i != 50; });
+	static_assert(ranges::RandomAccessRange<decltype(rng1)>);
+	CHECK_EQUAL(rng1, {0, 1, 2, 3, 4, 5, 6, 7, 8, 9});
+
+	/////////////////////////////////////////////////////////////////////////////////////////////////
+	//																															  //
+	// 						DISABLED because the lambda is not equality-preserving.						  //
+	//							Left in the review only to confirm that this is the case.					  //
+	//																															  //
+	/////////////////////////////////////////////////////////////////////////////////////////////////
+	// Check with a mutable predicate
+	// int rgi[] = {0,1,2,3,4,5,6,7,8,9};
+	// int cnt = 0;
+	// auto mutable_only = view::take_while(rgi, [cnt](int) mutable { return ++cnt <= 5;});
+	// CHECK_EQUAL(mutable_only, {0,1,2,3,4});
+	// static_assert(ranges::View<decltype(mutable_only)>);
+	// static_assert(!ranges::View<decltype(mutable_only) const>);
+
+	// Check with a mutable predicate
+	{
+		auto rng0 = view::iota(10) | view::take_while([](int i) mutable { return i != 25; });
```

Comment:
Ditto "should actually mutate something."

Reply:
Doing this means that the lambda isn't `const`-invocable, which causes waves for `const Pred*` in the sentinel. Would making it non-`const` pose any problems for [[res.on.data.races]](http://eel.is/c++draft/res.on.data.races)?

(This isn't a problem for `drop_while_view`, because it doesn't handle its predicate in the same way).

Reply:
If we allow non-`const`-invocable lambdas, the `mutable` lambda test should actually mutate something. If we forbid them, we can't test with a `mutable` lambda because it won't be `const`-invocable whether it actually mutates something or not.

Reply:
I think this is the only blocker now. I'm hesitant to forbid non-`const`-invocable lambdas, because it eliminates caches, and I feel like that's a good enough reason to go to the effort of allowing them (caches don't preclude equality-preservation, right?).

Reply:
Having said that, I can't think of any motivating uses for a cache, so maybe it belongs on the backlog for when it actually becomes necessary.

Reply:
I've been waffling back and forth on this after spending some time messing with `transform_view` yesterday. There's a tradeoff between "Do we want to make it easy to use correct non-`const`-invocable function objects" and "Do we want to make it hard to use incorrect non-`const`-invocable function objects." I doubt that users understand [res.on.data.races] or the consequences for their function objects that mutate internal state: if we simply accept non-`const`-invocable function objects and assume that the user is defending their internal state against data races, users *will* write bugs. If we only accept `const`-invocable function objects, then motivated users who *really* want mutable internal state will have to jump through hoops to write function objects with `mutable` members and more likely at least understand that they're doing something dangerous, if not necessarily *why* it's dangerous.